File: ts/packages/over-memory/src/transport.ts
import {
  parseEnvelope, EnvelopeKind, ClamatorTransportError,
  type Transport, type Dispatcher, type SendOptions,
} from '@clamator/protocol';
import type { MemoryBus } from './bus.js';

interface Pending {
  resolve: (env: Record<string, unknown>) => void;
  reject: (err: unknown) => void;
  timer: ReturnType<typeof setTimeout>;
}

export class MemoryTransport implements Transport {
  private state: 'idle' | 'started' | 'stopped' = 'idle';
  private pending = new Map<string, Pending>();
  private myServices = new Set<string>();

  constructor(private readonly bus: MemoryBus, private readonly _instanceId: string = 'mem') {}

  async registerService(name: string, dispatch: Dispatcher): Promise<void> {
    this.bus.register(name, dispatch);
    this.myServices.add(name);
  }

  async send(env: Record<string, unknown>, opts: SendOptions): Promise<Record<string, unknown>> {
    if (this.state !== 'started')
      throw new ClamatorTransportError(`transport not started (state=${this.state})`);
    const parsed = parseEnvelope(env);
    if (parsed.kind !== EnvelopeKind.Request)
      throw new ClamatorTransportError('send requires a request envelope');
    const dispatcher = this.bus.lookup(parsed.service);
    if (!dispatcher) {
      // Simulate the same -32601 path as a real adapter after delivery + lookup-failure on server.
      return {
        jsonrpc: '2.0', id: parsed.id,
        error: { code: -32601, message: 'Method not found', data: null },
      };
    }
    return await new Promise<Record<string, unknown>>((resolve, reject) => {
      const timer = setTimeout(() => {
        this.pending.delete(parsed.id as string);
        reject(new ClamatorTransportError('call timeout'));
      }, opts.timeoutMs);
      this.pending.set(String(parsed.id), { resolve, reject, timer });
      // Schedule async dispatch via microtask.
      queueMicrotask(async () => {
        try {
          const reply = await dispatcher(parsed);
          const p = this.pending.get(String(parsed.id));
          if (!p) return;
          this.pending.delete(String(parsed.id));
          clearTimeout(p.timer);
          if (reply === null) {
            p.reject(new ClamatorTransportError('dispatcher returned null for a request'));
          } else {
            p.resolve(reply);
          }
        } catch (e) {
          const p = this.pending.get(String(parsed.id));
          if (!p) return;
          this.pending.delete(String(parsed.id));
          clearTimeout(p.timer);
          p.reject(new ClamatorTransportError('dispatcher threw', e));
        }
      });
    });
  }

  async notify(env: Record<string, unknown>): Promise<void> {
    if (this.state !== 'started')
      throw new ClamatorTransportError(`transport not started (state=${this.state})`);
    const parsed = parseEnvelope(env);
    if (parsed.kind !== EnvelopeKind.Notification)
      throw new ClamatorTransportError('notify requires a notification envelope');
    const dispatcher = this.bus.lookup(parsed.service);
    if (!dispatcher) return;  // Silent drop, like a real fire-and-forget.
    queueMicrotask(() => { void dispatcher(parsed); });
  }

  async start(): Promise<void> {
    if (this.state === 'stopped') throw new ClamatorTransportError('transport has been stopped');
    this.state = 'started';
  }

  async stop(): Promise<void> {
    this.state = 'stopped';
    for (const [id, p] of this.pending.entries()) {
      clearTimeout(p.timer);
      p.reject(new ClamatorTransportError('transport stopped'));
      this.pending.delete(id);
    }
    for (const name of this.myServices) this.bus.unregister(name);
    this.myServices.clear();
  }
}

File: py/packages/over-memory/src/clamator_over_memory/transport.py
from __future__ import annotations

import asyncio
import logging
from typing import Any

from clamator_protocol import (
    ClamatorTransportError,
    Dispatcher,
    NotificationEnvelope,
    RequestEnvelope,
    build_error_response,
    parse_envelope,
)

from .bus import MemoryBus

logger = logging.getLogger(__name__)


class MemoryTransport:
    def __init__(self, bus: MemoryBus, instance_id: str = "mem") -> None:
        self._bus = bus
        self._instance_id = instance_id
        self._state: str = "idle"
        self._pending: dict[str, asyncio.Future[dict[str, Any]]] = {}
        self._my_services: set[str] = set()

    async def register_service(self, name: str, dispatch: Dispatcher) -> None:
        self._bus.register(name, dispatch)
        self._my_services.add(name)

    async def send(self, env: dict[str, Any], *, timeout: float) -> dict[str, Any]:
        if self._state != "started":
            raise ClamatorTransportError(f"transport not started (state={self._state})")
        parsed = parse_envelope(env)
        if not isinstance(parsed, RequestEnvelope):
            raise ClamatorTransportError("send requires a request envelope")
        dispatcher = self._bus.lookup(parsed.service)
        if dispatcher is None:
            return build_error_response(parsed.id, -32601, "Method not found")
        loop = asyncio.get_running_loop()
        fut: asyncio.Future[dict[str, Any]] = loop.create_future()
        self._pending[str(parsed.id)] = fut

        async def runner() -> None:
            try:
                reply = await dispatcher(parsed)
                if str(parsed.id) not in self._pending:
                    return
                self._pending.pop(str(parsed.id))
                if reply is None:
                    fut.set_exception(
                        ClamatorTransportError("dispatcher returned None for a request")
                    )
                else:
                    fut.set_result(reply)
            except Exception as e:  # noqa: BLE001
                logger.warning(
                    "dispatcher threw: service=%s",
                    parsed.service,
                    exc_info=True,
                    extra={"clamator": {"service": parsed.service, "rpc_id": str(parsed.id)}},
                )
                if str(parsed.id) in self._pending:
                    self._pending.pop(str(parsed.id))
                    if not fut.done():
                        fut.set_exception(ClamatorTransportError("dispatcher threw", cause=e))

        loop.create_task(runner())
        try:
            return await asyncio.wait_for(fut, timeout=timeout)
        except TimeoutError as e:
            self._pending.pop(str(parsed.id), None)
            raise ClamatorTransportError("call timeout") from e

    async def notify(self, env: dict[str, Any]) -> None:
        if self._state != "started":
            raise ClamatorTransportError(f"transport not started (state={self._state})")
        parsed = parse_envelope(env)
        if not isinstance(parsed, NotificationEnvelope):
            raise ClamatorTransportError("notify requires a notification envelope")
        dispatcher = self._bus.lookup(parsed.service)
        if dispatcher is None:
            return
        asyncio.get_running_loop().create_task(dispatcher(parsed))

    async def start(self) -> None:
        if self._state == "stopped":
            raise ClamatorTransportError("transport has been stopped")
        self._state = "started"

    async def stop(self) -> None:
        self._state = "stopped"
        for fut in list(self._pending.values()):
            if not fut.done():
                fut.set_exception(ClamatorTransportError("transport stopped"))
        self._pending.clear()
        for name in list(self._my_services):
            self._bus.unregister(name)
        self._my_services.clear()
